installShellFiles: rework and add installBin function

AndersonTorres opened the pull request:
## Description of changes

nawk for testing purposes

The rationales of this PR are better understood by reading the commit messages, but an executive resume is:

- migrate to by-name
- create a test directory instead of a test file
- new function installBin

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

AndersonTorres commented:
@afh I have taken your bump to make a test here. Can you review it?

AndersonTorres commented:
Hum, does it cause mass-rebuild?

AndersonTorres committed:
CODEOWNERS: update ownership for installShellFiles

Since Ericson2314 is the original codeowner of the files.

AndersonTorres committed:
doc: rewrite installShellFiles

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4382

afh reviewed (approved):
Thanks for this, @AndersonTorres, I think the addition of `installBin` makes much sense 👍 
Several packages that I've come across  install their binaries differently, mostly `mkdir` with `cp` or `mv` or `install`. This PR would help to consolidate this.

I haven't had much exposure to nixpkgs internals so please take this approval with a grain of salt 🙂

AndersonTorres commented:
Two extra ideas:
- rename this to `installArtifacts`,  since it is a more general name than `installShellFiles`
  - green alias
- include other artifact installers like `installDoc`

philiptaron commented on pkgs/by-name/in/installShellFiles/package.nix:
```diff
@@ -0,0 +1,28 @@
+in
+setupHook.overrideAttrs (oldAttrs: {
+  passthru = (oldAttrs.passthru or { }) // {
+    tests = lib.packagesFromDirectoryRecursive {
+      inherit callPackage;
+      directory = ./tests;
+    };
```
I _love_ this construction. We should do this more often.

philiptaron commented on pkgs/by-name/in/installShellFiles/setup-hook.sh:
```diff
@@ -233,3 +233,26 @@ installShellCompletion() {
+installBin() {
+    local path
+    for path in "$@"; do
+        if (( "${NIX_DEBUG:-0}" >= 1 )); then
+            echo "installBin: installing $path"
+        fi
```
As of #331560, this is `nixWarnLog "installBin: installing $path"`. Can also use the following:

* nixChattyLog
* nixDebugLog
* nixErrorLog (this is always)
* nixInfoLog
* nixNoticeLog
* nixTalkativeLog
* nixVomitLog
* nixWarnLog

AndersonTorres replied:
Can I use `nixInfoLog` instead?
It does not look like a warning to me.

philiptaron replied:
Please do. My intent in introducing named functions was to make these "what level" discussions more vibrant and useful.

AndersonTorres replied:
P.S.: can I wait for the merge into Master branch? I usually test things at Master to not deal with mass rebuilds.

philiptaron replied:
Merge happened.
- #332764

philiptaron commented on pkgs/by-name/in/installShellFiles/setup-hook.sh:
```diff
@@ -233,3 +233,26 @@ installShellCompletion() {
+        local outRoot
+        outRoot=${!outputBin:?}
+
+        install -Dm755 -T "$path" "${outRoot}/bin/$basename" || return
```
I don't see this "succeed on error" on other things. Why this one?

AndersonTorres replied:
Well, this is the same construction on `installManPage`:

https://github.com/NixOS/nixpkgs/pull/332612/files/ce45e04e92ce8f22fb281db1d9d2c9a84bf98de5#diff-1bf14a6bd9bd0976c8094adb934eefa32d7904ce6ef119a7dae5cd6e8fbfc63eL48

philiptaron replied:
It was introduced in https://github.com/NixOS/nixpkgs/commit/43dade238f39f. @matthewbauer had some commentary on that commit.

Note that `installManPage` is the only one with that questionable pattern. I'm pretty against this error swallowing behavior. There's no rationale in the original commit, and it's liable to look right and be wrong, which is ... not great.

AndersonTorres replied:
I rewrote the functions to be more conformant.

philiptaron commented on pkgs/by-name/in/installShellFiles/package.nix:
```diff
@@ -0,0 +1,28 @@
+
+let
+  setupHook = makeSetupHook { name = "install-shell-files"; } (
+    builtins.path {
+      name = "install-shell-files-hook";
+      path = ./setup-hook.sh;
+      recursive = false;
+      sha256 = "sha256-cLDb6u/2p82wecm4P9spdSDYvBWQv96DnUKXXdzAeYE=";
+    }
```
I'm a bit leery of this `builtins.path` call. Why add it here, when the original didn't include it?

AndersonTorres replied:
It has some purposes.
- Reproducibility, as explained in https://nix.dev/guides/best-practices.html#reproducible-source-paths
- the `name` attribute can be useful to track it in the Nix Store and in the logs
- the hash assures the integrity of this file

I think it is a nice add-on for such an important file.

AndersonTorres replied:
I will remove.
Too much maintenance work.

philiptaron commented on pkgs/by-name/in/installShellFiles/tests/install-bin-output.nix:
For this and other tests, consider using `testers.testEqualContents` rather than doing positive and negative tests directly. Since that tester compares two directories with diffoscope, it covers:

1. mode bits
2. contents
3. filenames

in a thorough way.

Some of these tests are super simple and it's not needed per se. But I really like it as a way to make sure that a given hook does exactly what it's supposed to do.

AndersonTorres replied:
I kept the previous testing routines as is, just scattering them along many files.

Certainly I can convert them right now.

SuperSandro2000 replied:
I mean if we merely move the tests around, what is the goal of this PR? Adding new tests and making them simpler and more awesome would be a good goal.

AndersonTorres replied:
> I mean if we merely move the tests around, what is the goal of this PR? 

This PR, first and foremost, creates a new function, `installBin`, to install files into `outputBin`.

Second, given that per RFC 140 the old hierarchy becomes legacy/obsolete and a more self-contained approach is encouraged, this PR moves the tests to the same directory tree that the package itself belongs.

Third, `packagesFromDirectoryRecursive` is perfect for this job of creating tests: instead of throwing them in a same text file, it is way cleaner to put them in the same directory.

philiptaron reviewed (changes requested):
Requesting changes for the following:

1. Explanation and ideally removal of the `builtins.path` addition.
2. Explanation and ideally removal of the `|| return` in `installBin`.

I'd also love to see the following changes, but they do not block the PR:

1. Use of `testers.testEqualContents` as a way to validate that the directory structure produced looks _exactly_ a certain way.
2. Use of a `nix*Log` function instead of a direct check of the `$NIX_DEBUG` environment variable.
3. Use of `makeSetupHook`'s `passthru` parameter rather than `overrideAttrs`.

AndersonTorres commented on pkgs/by-name/in/installShellFiles/tests/install-manpage.nix:
```diff
@@ -0,0 +1,23 @@
+runCommandLocal "install-shell-files--install-manpage"
+  {
+    nativeBuildInputs = [ installShellFiles ];
+    meta.platforms = lib.platforms.all;
```
I merely kept what is there before.

SuperSandro2000 replied:
Probably to also build them for darwin.

AndersonTorres committed:
installShellFiles: move setup script to the same directory

So that it can be migrated to by-name later.

AndersonTorres committed:
installShellFiles: migrate tests to tests subdirectory

Rationale: Since RFCs 140 and 146, the old category-based hierarchy is
deprecated and obsolete, and a new approach took place: packages should be as
self-contained as possible.

This paradigm is reflected in many new strict checks that prohibit a package to
refer to files outside its directory tree.

Following this spirit, this commit essentially moves nixpkgs
pkgs/test/default.nix to ./tests/default.nix.

Further, to keep the top-level `tests` attribute, a green alias is kept in the
place of older file.

AndersonTorres committed:
installShellFiles: nixfmt-rfc-style

AndersonTorres committed:
installShellFiles: migrate to by-name

AndersonTorres committed:
installShellFiles: dismember tests into their own files

It makes the maintenance and creation of tests more intuitive: just throw a Nix
package under tests!

AndersonTorres committed:
installShellFiles: add new function installBin

AndersonTorres committed:
installShellFiles: add new tests install-bin and install-bin-output

AndersonTorres commented on pkgs/by-name/in/installShellFiles/package.nix:
```diff
@@ -0,0 +1,28 @@
+# - setup-hook.sh (with usage examples)
+
+let
+  setupHook = makeSetupHook { name = "install-shell-files"; } (
```
I fear this will not be possible. By any strange reason it generated some failure.

philiptaron replied:
What's the error?

AndersonTorres replied:
I will try to reproduce it.

philiptaron commented on pkgs/by-name/in/installShellFiles/setup-hook.sh:
```diff
@@ -188,7 +186,7 @@ installShellCompletion() {
             *)
                 # Our list of shells is out of sync with the flags we accept or extensions we detect.
-                echo 'installShellCompletion: internal error' >&2
+                nixErrorLog '${FUNCNAME[0]}: internal error'
```
`'` means no expansion; I think you mean `"`

philiptaron commented on pkgs/by-name/in/installShellFiles/setup-hook.sh:
```diff
@@ -206,7 +204,7 @@ installShellCompletion() {
         *)
             # Our list of shells is out of sync with the flags we accept or extensions we detect.
-            echo 'installShellCompletion: internal error' >&2
+            nixErrorLog '${FUNCNAME[0]}: internal error'
```
Same single quote business

AndersonTorres replied:
_D'oh!_

philiptaron commented on pkgs/by-name/in/installShellFiles/setup-hook.sh:
```diff
@@ -218,18 +216,44 @@ installShellCompletion() {
 
         if [ ! -s "$outPath" ]; then
-            echo "installShellCompletion: error: installed shell completion file \`$outPath' does not exist or has zero size" >&2
+            nixErrorLog "${FUNCNAME[0]}:" "installed shell completion file \`$outPath' does not exist or has zero size"
```
Curious why not just the following:
```suggestion
            nixErrorLog "${FUNCNAME[0]}: installed shell completion file \`$outPath' does not exist or has zero size"
```

AndersonTorres replied:
`echoCmd` muscular memeory

philiptaron commented on pkgs/by-name/in/installShellFiles/setup-hook.sh:
```diff
@@ -45,7 +43,9 @@ installManPage() {
         fi
-        install -Dm644 -T "$path" "${outRoot}/share/man/man$suffix/$basename" || return
+        local outPath="${outRoot}/share/man/man$suffix/$basename"
+        install -Dm644 -T "$path" "$outPath" || \
+            { nixErrorLog "${FUNCNAME[0]}:" "install $path to $outPath failed"; return 1; }
```
Curious why not the following:
```suggestion
        if ! install -Dm644 -T "$path" "$outPath"; then
            nixErrorLog "${FUNCNAME[0]}: install $path to $outPath failed"
            return 1
        fi
            
```

Alternatively, let the `install` tool output its errors: https://git.savannah.gnu.org/cgit/coreutils.git/tree/src/install.c shows that it's got substantial error logging.

AndersonTorres replied:
Good.

philiptaron reviewed (commented):
A few thought skimming this. I haven't tested it yet.

SuperSandro2000 commented on pkgs/test/install-shell-files/default.nix:
> Further, to keep the top-level `tests` attribute, a green alias is kept in the
place of older file.

Why not just update that instead of keeping this almost empty file?

AndersonTorres replied:
The `tests` top-level attribute contains multiple other packages.

SuperSandro2000 commented on pkgs/by-name/in/installShellFiles/setup-hook.sh:
```diff
@@ -233,3 +233,27 @@ installShellCompletion() {
+# installBin <path> [...<path>]
+#
+# Install each argument to $outputBin
+# Heavily inspired in installManPage above
```
```suggestion
```
they will diverge over time

SuperSandro2000 commented on .github/CODEOWNERS:
This commit can be combined with the last one

SuperSandro2000 commented on doc/hooks/installShellFiles.section.md:
```diff
@@ -1,16 +1,72 @@
+{
+  nativeBuildInputs = [ installShellFiles ];
+  postInstall = ''
+    mv fromsea.3 delmar.3
```
I don't understand the purpose of this line.

AndersonTorres replied:
To simulate cases where the manpage name is not suitable. E. G. nawk names its manpage as `awk.1`.

AndersonTorres replied:
I will comment.

SuperSandro2000 commented on doc/hooks/installShellFiles.section.md:
```diff
@@ -1,16 +1,72 @@
+{
+  nativeBuildInputs = [ installShellFiles ];
+  postInstall = ''
+    mv fromsea delmar
```
same here

AndersonTorres committed:
installShellFiles: rewrite functions

- Use FUNCNAME to track the name of function being called
- use long options for install commands
- use nix*Log functions for logging, instead of custom echoes
- remove silent error construction `|| return`

AndersonTorres committed:
nawk: uses installBin to install the binary

AndersonTorres committed:
installShellFiles: set passthru directly instead of overriding it

philiptaron reviewed (approved):
I intend to merge after doing a set of sample builds.

AndersonTorres commented:
@philiptaron you commented about using `testEqualContent`. However, from what I have seen, it does test singular files, not directory trees. So I prefer to keep it as is.

@infinisil i want to ask your opinion about a specific decision.

As I have written in https://github.com/NixOS/nixpkgs/pull/332612/commits/f72e74d7aef142ea16d0841fa4322fc1310bb913:

> Rationale: Since RFCs 140 and 146, the old category-based hierarchy is
> deprecated and obsolete, and a new approach took place: packages should be as
> self-contained as possible.
> 
> This paradigm is reflected in many new strict checks that prohibit a package to
> refer to files outside its directory tree.
> 
> Following this spirit, this commit essentially moves nixpkgs
> pkgs/test/default.nix to ./tests/default.nix.

What do you think? Is this a good approach?

infinisil commented:
> What do you think? Is this a good approach?

That looks great to me, nice work!

infinisil commented on pkgs/by-name/in/installShellFiles/package.nix:
```diff
@@ -0,0 +1,16 @@
+# See the header comment in ./setup-hook.sh for example usage.
+makeSetupHook {
+  name = "install-shell-files";
+  passthru = {
+    tests = lib.packagesFromDirectoryRecursive {
+      inherit callPackage;
+      directory = ./tests;
+    };
```
This is not worth addressing here, but note that this has a minor problem: Due to `packagesFromDirectoryRecursive` recursing for directories, a standard `nix-build -A installShellFiles.tests` wouldn't run any nested ones, because `nix-build` doesn't recurse more than one level. In my experience, recursive attribute sets give rise to various such oddities, which is why I recommend avoiding them. Cc @philiptaron

philiptaron replied:
Good to know! There is likely need then for a simple non-recursive variant for these flat use cases.

infinisil reviewed (approved):
This is well done commit history! I very briefly looked through the changes, LGTM!
